Add support for multiple operators via several improvements

lisa opened the pull request:
This PR makes the following changes:

# scripts/gen_operator_csv.py (8fd92c1c88adc2cae821422e2c669554f658641e)

* Fix typo, add guardrails around missing assets
* If the `manifests` directory is missing, complain and quit. If there are no `Deployments` detected, complain and quit.

# Makefile (4d99ecdc6fa2fc7b642363e4a9dddfdc04fc8e8e)

* In all places substitute filesystem glob matching in `$(SOURCE_DIR)/**/` to iterate over `$(OPERATORS)` and determine the name of the repository from the list.
* Adds guardrails around certain variables that become part of the running environment due to `eval $(make -C operatordir) env`:
  * Tests are made to ensure the variables the make target cares about are found
  * Unsets variables so that subsequent loop iterations (and other make targets) aren't polluted with previous values
* Makes more explicit the testing around failures: Instead of `|| (... && exit)`, check the previous command's exit value in an `if-then-else`-style control block.
* Created an `env` target, which echos nothing, so that `functions.mk`'s `/process_template` doesn't emit an error.
* Most targets are now silent; they do not print the Bash that make is running. (To be more verbose try `make SHELL="bash -x"`)
* Consolidated multiple `mkdir -p manifests` into its own make target, `manifestdir`, with accompanying variable `MANIFESTDIR`

lisa committed:
Fix typo, add guardrails around missing assets

If the `manifests` directory is missing, complain and quit. If there are
no `Deployments` detected, complain and quit.

Signed-off-by: Lisa Seelye <lseelye@redhat.com>

lisa committed:
Harden and add support for configure-alertmanager-operator

This commit makes the following changes:

* In all places substitute filesystem glob matching in `$(SOURCE_DIR)/**/` to iterate over `$(OPERATORS)` and determine the name of the repository from the list.
* Adds guardrails around certain variables that become part of the running environment due to `eval $(make -C operatordir) env`:
  * Tests are made to ensure the variables the make target cares about are found
  * Unsets variables so that subsequent loop iterations (and other make targets) aren't polluted with previous values
* Makes more explicit the testing around failures: Instead of `|| (... && exit)`, check the previous command's exit value in an `if-then-else`-style control block.
* Created an `env` target, which echos nothing, so that `functions.mk`'s `/process_template` doesn't emit an error.
* Most targets are now silent; they do not print the Bash that make is running. (To be more verbose try `make SHELL="bash -x"`)
* Consolidated multiple `mkdir -p manifests` into its own make target, `manifestdir`, with accompanying variable `MANIFESTDIR`

Signed-off-by: Lisa Seelye <lseelye@redhat.com>

jewzaam commented on Makefile:
```diff
@@ -39,9 +39,10 @@ GIT_TAG=release-$(CATALOG_VERSION)
 # List of github.org repositories containing operators
 # This is in the format username/reponame separated by space:  user1/repo1 user2/repo2 user3/repo3
-OPERATORS := openshift/dedicated-admin-operator
+OPERATORS := openshift/dedicated-admin-operator dak1n1/configure-alertmanager-operator
```
```suggestion
OPERATORS := openshift/dedicated-admin-operator openshift/configure-alertmanager-operator
```

jewzaam replied:
If we don't have this in the openshift org it needs created.  This is a requirement of the operator work. FYI @dak1n1

lisa replied:
Whoops, you're right. The repo exists, but she's still curating her PR.

jewzaam commented on Makefile:
```diff
@@ -60,31 +61,30 @@ clean:
 
 .PHONY: manifests
-manifests: manifests/catalog manifests/operators
+manifests: manifestdir manifests/catalog manifests/operators 
```
```suggestion
manifests: manifestdir manifests/catalog manifests/operators
```

jewzaam commented on Makefile:
```diff
@@ -93,17 +93,50 @@ operator-source:
+		operator="$$(echo $$operatorrepo | cut -d / -f2)" ;\
+		echo "Building catalog for $$operator in $(SOURCE_DIR)/$$operator" ;\
+		eval $$($(MAKE) -C $(SOURCE_DIR)/$$operator env --no-print-directory); \
+		if [[ -z "$${OPERATOR_NAME}" || -z "$${OPERATOR_NAMESPACE}" || -z "$${OPERATOR_VERSION}" ]]; then \
```
```suggestion
		if [[ -z "$${OPERATOR_NAME}" || -z "$${OPERATOR_NAMESPACE}" || -z "$${OPERATOR_VERSION}" || -z "$${OPERATOR_IMAGE_URI}" ]]; then \
```

jewzaam commented on Makefile:
```diff
@@ -93,17 +93,50 @@ operator-source:
+		echo "Building catalog for $$operator in $(SOURCE_DIR)/$$operator" ;\
+		eval $$($(MAKE) -C $(SOURCE_DIR)/$$operator env --no-print-directory); \
+		if [[ -z "$${OPERATOR_NAME}" || -z "$${OPERATOR_NAMESPACE}" || -z "$${OPERATOR_VERSION}" ]]; then \
+			echo "Couldn't determine OPERATOR_NAME, OPERATOR_NAMESPACE or OPERATOR_VERSION from $(SOURCE_DIR)/$$operator. make env output follows" ; \
```
```suggestion
			echo "Couldn't determine OPERATOR_NAME, OPERATOR_NAMESPACE, OPERATOR_VERSION or OPERATOR_IMAGE_URI from $(SOURCE_DIR)/$$operator. make env output follows" ; \
```

jewzaam commented on Makefile:
```diff
@@ -93,17 +93,50 @@ operator-source:
+			unset OPERATOR_NAME OPERATOR_VERSION OPERATOR_NAMESPACE OPERATOR_IMAGE_URI ;\
+			exit 3 ;\
+		else \
+			./scripts/gen_operator_csv.py $(SOURCE_DIR)/$$operator $$OPERATOR_NAME $$OPERATOR_NAMESPACE $$OPERATOR_VERSION $(IMAGE_REGISTRY)/$(IMAGE_REPOSITORY)/$$OPERATOR_NAME:v$$OPERATOR_VERSION $(CHANNEL) 1>/dev/null ;\
```
```suggestion
			./scripts/gen_operator_csv.py $(SOURCE_DIR)/$$operator $$OPERATOR_NAME $$OPERATOR_NAMESPACE $$OPERATOR_VERSION $$OPERATOR_IMAGE_URI $(CHANNEL) 1>/dev/null ;\
```

jewzaam commented on Makefile:
```diff
@@ -93,17 +93,50 @@ operator-source:
+			if [[ $$? -ne 0 ]]; then \
+				echo "Failed to generate, cleaning up catalog-manifests/$$OPERATOR_NAME/$$OPERATOR_VERSION" ;\
+				rm -rf catalog-manifests/$$OPERATOR_NAME/$$OPERATOR_VERSION ;\
+				unset OPERATOR_NAME OPERATOR_VERSION OPERATOR_NAMESPACE OPERATOR_IMAGE_URI ;\
```
Not needed if about to `exit 3`

lisa replied:
I know, but for completeness sake 🤷‍♀️

jewzaam commented on Makefile:
```diff
@@ -93,17 +93,50 @@ operator-source:
+				unset OPERATOR_NAME OPERATOR_VERSION OPERATOR_NAMESPACE OPERATOR_IMAGE_URI ;\
+				exit 3; \
+			fi ;\
+			unset OPERATOR_NAME OPERATOR_VERSION OPERATOR_NAMESPACE OPERATOR_IMAGE_URI ;\
```
Lots of unset.  Suggest moving this to the beginning of each loop iteration.  It only needs done when initializing this loop.  If a cleanup is desired, one outside the loop execution is fine too.

jewzaam commented on Makefile:
```diff
@@ -93,17 +93,50 @@ operator-source:
-	for DIR in $(SOURCE_DIR)/**/; do \
-		eval $$($(MAKE) -C $$DIR env --no-print-directory); \
-		docker pull $(IMAGE_REGISTRY)/$(IMAGE_REPOSITORY)/$$OPERATOR_NAME:v$$OPERATOR_VERSION || (echo "Image cannot be pulled: $(IMAGE_REGISTRY)/$(IMAGE_REPOSITORY)/$$OPERATOR_NAME:v$$OPERATOR_VERSION" && exit 1); \
+	@unset OPERATOR_NAME OPERATOR_VERSION OPERATOR_NAMESPACE OPERATOR_IMAGE_URI ;\
```
See comments and suggested changes from the `catalog` target.

jewzaam committed:
Update Makefile

remove extra whitespace

Co-Authored-By: lisa <18159+lisa@users.noreply.github.com>

jewzaam committed:
Update Makefile

Co-Authored-By: lisa <18159+lisa@users.noreply.github.com>

jewzaam committed:
Apply suggestions from code review

Require the presence of `OPERATOR_IMAGE_URI` in operator make files, and also use it to call `scripts/gen_operator_csv.py`

Co-Authored-By: lisa <18159+lisa@users.noreply.github.com>

lisa committed:
unset variables at the top of the loop

Signed-off-by: Lisa Seelye <lseelye@redhat.com>

jewzaam commented on Makefile:
```diff
@@ -93,17 +93,44 @@ operator-source:
+		org="$$(echo $$operator | cut -d / -f 1)" ; \
+		reponame="$$(echo $$operator | cut -d / -f 2-)" ; \
+		eval $$($(MAKE) -C $(SOURCE_DIR)/$$reponame env --no-print-directory); \
+		if [[ -z "$${OPERATOR_NAME}" || -z "$${OPERATOR_NAMESPACE}" || -z "$${OPERATOR_VERSION}" ]]; then \
```
```suggestion
		if [[ -z "$${OPERATOR_NAME}" || -z "$${OPERATOR_NAMESPACE}" || -z "$${OPERATOR_VERSION}" || -z "$${OPERATOR_IMAGE_URI}" ]]; then \
```

jewzaam commented on Makefile:
```diff
@@ -93,17 +93,44 @@ operator-source:
+		reponame="$$(echo $$operator | cut -d / -f 2-)" ; \
+		eval $$($(MAKE) -C $(SOURCE_DIR)/$$reponame env --no-print-directory); \
+		if [[ -z "$${OPERATOR_NAME}" || -z "$${OPERATOR_NAMESPACE}" || -z "$${OPERATOR_VERSION}" ]]; then \
+			echo "Couldn't determine OPERATOR_NAME, OPERATOR_NAMESPACE or OPERATOR_VERSION from $(SOURCE_DIR)/$$operator. make env output follows" ; \
```
```suggestion
			echo "Couldn't determine OPERATOR_NAME, OPERATOR_NAMESPACE, OPERATOR_VERSION or OPERATOR_IMAGE_URI from $(SOURCE_DIR)/$$operator. make env output follows" ; \
```

jewzaam commented on Makefile:
```diff
@@ -93,17 +93,44 @@ operator-source:
+			$(MAKE) -C $(SOURCE_DIR)/$$operator env ; \
+			exit 3 ;\
+		else \
+			docker pull $(IMAGE_REGISTRY)/$(IMAGE_REPOSITORY)/$$OPERATOR_NAME:v$$OPERATOR_VERSION ;\
```
```suggestion
			docker pull $$OPERATOR_IMAGE_URI ;\
```

jewzaam committed:
Updated README for OPERATOR_IMAGE_URI

jewzaam commented:
PR into this PR to update readme: https://github.com/lisa/osd-operators-registry/pull/1

lisa committed:
Merge pull request #1 from jewzaam/lisa-readme

Updated README for OPERATOR_IMAGE_URI

jewzaam committed:
Update Makefile

Co-Authored-By: lisa <18159+lisa@users.noreply.github.com>

jewzaam committed:
Update Makefile

Co-Authored-By: lisa <18159+lisa@users.noreply.github.com>

jewzaam committed:
Update Makefile

Co-Authored-By: lisa <18159+lisa@users.noreply.github.com>

jewzaam commented:
Changes in this repo are good, so:
/approve

But the config-alertmanager-operator repo has problems, noting here:

- inconsistent make target for build
- missing make target for push
- manifest filenames not catalogsource friendly (format is <text>.<kind.lowerCase>.yaml)

Bottom line is merging this breaks generation of the catalog.  Need the operator to be updated before we can merge. So:
/hold

jewzaam commented:
Filenames can be handled here.  I'll put in a quick PR for that.  See https://github.com/openshift/osd-operators-registry/issues/15

jewzaam commented:
Filename renames here: https://github.com/openshift/osd-operators-registry/pull/16

lisa commented:
This will conflict with #17.

lisa committed:
Remove alertmanager operator from operator list

This will effectively change #12 into two:

* Improvements that support adding alertmanager
* adding alertmanager

Part two will take place in another PR once alertmanager operator is in its master branch

jewzaam commented on Makefile:
```diff
@@ -93,17 +93,44 @@ operator-source:
+		else \
+			docker pull $$OPERATOR_IMAGE_URI ;\
+			if [[ $$? -ne 0 ]]; then \
+				echo "Image cannot be pulled: $(IMAGE_REGISTRY)/$(IMAGE_REPOSITORY)/$$OPERATOR_NAME:v$$OPERATOR_VERSION" ;\
```
```suggestion
				echo "Image cannot be pulled: $$OPERATOR_IMAGE_URI ;\
```

rogbas commented:
/lgtm

jewzaam committed:
Update Makefile

Co-Authored-By: lisa <18159+lisa@users.noreply.github.com>

jewzaam commented:
/lgtm
